File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/picker_dialog_month/MonthPickerDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_month;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment.DialogCallbacks;
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom.CustomDialogBuilder;

import java.util.ArrayList;

import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LifecycleOwner;

public class MonthPickerDialogBuilder extends BaseDialogBuilder<MonthPickerDialog> {

    private MonthPickerDialog.SelectionChangedListener<int[]> monthSetListener;
    private ArrayList<int[]> disabledMonths;
    private ArrayList<int[]> enabledMonths;
    private String dateFormat;
    private int[] initialSelection;
    private int minYear = 1970;
    private int maxYear = 2100;

    public MonthPickerDialogBuilder(FragmentManager manager, String dialogTag) {
        super(manager, dialogTag);
    }

    public MonthPickerDialogBuilder(LifecycleOwner lifecycleOwner, FragmentManager manager, String dialogTag) {
        super(lifecycleOwner, manager, dialogTag);
    }

    @Override
    public MonthPickerDialogBuilder withTitle(String title) {
        return (MonthPickerDialogBuilder) super.withTitle(title);
    }

    @Override
    public MonthPickerDialogBuilder withMessage(String message) {
        return (MonthPickerDialogBuilder) super.withMessage(message);
    }

    @Override
    public MonthPickerDialogBuilder withPositiveButton(BaseDialogFragment.DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (MonthPickerDialogBuilder) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public MonthPickerDialogBuilder withNegativeButton(BaseDialogFragment.DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (MonthPickerDialogBuilder) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public MonthPickerDialogBuilder withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        return (MonthPickerDialogBuilder) super.withOnCancelListener(listener);
    }

    @Override
    public MonthPickerDialogBuilder withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        return (MonthPickerDialogBuilder) super.withOnShowListener(listener);
    }

    @Override
    public MonthPickerDialogBuilder withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        return (MonthPickerDialogBuilder) super.withOnHideListener(listener);
    }

    @Override
    public MonthPickerDialogBuilder withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (MonthPickerDialogBuilder) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public MonthPickerDialogBuilder withDialogType(BaseDialogFragment.DialogTypes dialogType) {
        return (MonthPickerDialogBuilder) super.withDialogType(dialogType);
    }

    @Override
    public MonthPickerDialogBuilder withAnimations(BaseDialogFragment.DialogAnimationTypes animation) {
        return (MonthPickerDialogBuilder) super.withAnimations(animation);
    }

    @Override
    public MonthPickerDialogBuilder withDialogCallbacks(DialogCallbacks callbacks) {
        return (MonthPickerDialogBuilder) super.withDialogCallbacks(callbacks);
    }

    public MonthPickerDialogBuilder withDisabledMonths(ArrayList<int[]> disabledMonths) {
        this.disabledMonths = disabledMonths;
        return this;
    }

    public MonthPickerDialogBuilder withEnabledMonths(ArrayList<int[]> enabledMonths) {
        this.enabledMonths = enabledMonths;
        return this;
    }

    public MonthPickerDialogBuilder withOnDateSelectedEvent(MonthPickerDialog.SelectionChangedListener<int[]> listener) {
        this.monthSetListener = listener;
        return this;
    }

    public MonthPickerDialogBuilder withSelection(int year, int month) {
        this.initialSelection = new int[]{year, month};
        return this;
    }

    public MonthPickerDialogBuilder withMinYear(int year) {
        this.minYear = year;
        return this;
    }

    public MonthPickerDialogBuilder withMaxYear(int year) {
        this.maxYear = year;
        return this;
    }

    public MonthPickerDialogBuilder withDateFormat(String dateFormat) {
        this.dateFormat = dateFormat;
        return this;
    }

    @Override
    public MonthPickerDialog buildDialog() {
        MonthPickerDialog dialogFragment = (MonthPickerDialog) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = MonthPickerDialog.newInstance(positiveButtonConfiguration, negativeButtonConfiguration, dateFormat, cancelableOnClickOutside, animation);
        dialogFragment.setLifecycleOwner(dialogLifecycleOwner);
        dialogFragment.setDialogCallbacks(dialogCallbacks);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener);
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        dialogFragment.setOnSelectionChangedListener(monthSetListener);
        dialogFragment.setDisabledMonths(disabledMonths);
        dialogFragment.setEnabledMonths(enabledMonths);
        dialogFragment.setCalendarBounds(minYear, maxYear);
        if (initialSelection != null)
            dialogFragment.setSelection(initialSelection[0], initialSelection[1]);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/picker_dialog_month/MonthPickerDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_month

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.BasePickerDialogFragment.SelectionChangedListener
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.*
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*
import com.github.rooneyandshadows.lightbulb.dialogs.picker_dialog_month.MonthPickerDialog.*

@Suppress("unused")
class MonthPickerDialogBuilder @JvmOverloads constructor(
    dialogTag: String,
    dialogParentFragmentManager: FragmentManager,
    dialogLifecycleOwner: LifecycleOwner? = null,
    initialDialogState: Bundle? = null
) : BaseDialogBuilder<MonthPickerDialog>(dialogTag, dialogParentFragmentManager, dialogLifecycleOwner, initialDialogState) {
    private var monthSetListener: SelectionChangedListener<Month>? = null
    private var disabledMonths: MutableList<Month> = mutableListOf()
    private var enabledMonths: MutableList<Month> = mutableListOf()
    private var dateFormat: String? = null
    private var initialSelection: Month? = null
    private var minYear = 1970
    private var maxYear = 2100

    @JvmOverloads
    constructor(dialogTag: String, fragment: Fragment, initialDialogState: Bundle? = null) : this(
        dialogTag,
        fragment.childFragmentManager,
        fragment,
        initialDialogState
    )

    @JvmOverloads
    constructor(dialogTag: String, activity: FragmentActivity, initialDialogState: Bundle? = null) : this(
        dialogTag,
        activity.supportFragmentManager,
        activity,
        initialDialogState
    )

    @Override
    override fun setupNonRetainableSettings(dialog: MonthPickerDialog) {
        dialog.apply {
            monthSetListener?.apply { addOnSelectionChangedListener(this) }
        }
    }

    @Override
    override fun setupRetainableSettings(dialog: MonthPickerDialog) {
        dialog.apply {
            val builder = this@MonthPickerDialogBuilder
            setDisabledMonths(builder.disabledMonths)
            setEnabledMonths(builder.enabledMonths)
            setDialogDateFormat(dateFormat)
            setCalendarBounds(minYear, maxYear)
            setSelection(initialSelection)
        }
    }

    @Override
    override fun initializeNewDialog(): MonthPickerDialog {
        return MonthPickerDialog.newInstance()
    }

    @Override
    override fun withTitle(title: String): MonthPickerDialogBuilder {
        return super.withTitle(title) as MonthPickerDialogBuilder
    }

    @Override
    override fun withMessage(message: String): MonthPickerDialogBuilder {
        return super.withMessage(message) as MonthPickerDialogBuilder
    }

    @Override
    override fun withButton(configuration: DialogButton): MonthPickerDialogBuilder {
        return super.withButton(configuration) as MonthPickerDialogBuilder
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): MonthPickerDialogBuilder {
        return super.withOnCancelListener(listener) as MonthPickerDialogBuilder
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): MonthPickerDialogBuilder {
        return super.withOnShowListener(listener) as MonthPickerDialogBuilder
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): MonthPickerDialogBuilder {
        return super.withOnHideListener(listener) as MonthPickerDialogBuilder
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): MonthPickerDialogBuilder {
        return super.withCancelOnClickOutside(closeOnClickOutside) as MonthPickerDialogBuilder
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): MonthPickerDialogBuilder {
        return super.withDialogType(dialogType) as MonthPickerDialogBuilder
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): MonthPickerDialogBuilder {
        return super.withAnimations(animation) as MonthPickerDialogBuilder
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): MonthPickerDialogBuilder {
        return super.withDialogListeners(listeners) as MonthPickerDialogBuilder
    }

    fun withDisabledMonths(disabledMonths: List<Month>): MonthPickerDialogBuilder {
        this.disabledMonths.apply {
            clear()
            addAll(disabledMonths)
        }
        return this
    }

    fun withEnabledMonths(enabledMonths: List<Month>): MonthPickerDialogBuilder {
        this.enabledMonths.apply {
            clear()
            addAll(enabledMonths)
        }
        return this
    }

    fun withOnDateSelectedEvent(listener: SelectionChangedListener<Month>): MonthPickerDialogBuilder {
        monthSetListener = listener
        return this
    }

    fun withSelection(month: Month): MonthPickerDialogBuilder {
        initialSelection = month
        return this
    }

    fun withSelection(year: Int, month: Int): MonthPickerDialogBuilder {
        initialSelection = Month(year, month)
        return this
    }

    fun withMinYear(year: Int): MonthPickerDialogBuilder {
        minYear = year
        return this
    }

    fun withMaxYear(year: Int): MonthPickerDialogBuilder {
        maxYear = year
        return this
    }
}